### PR TITLE
refactored functions to read and write binaries

### PR DESCRIPTION
**Pull request by czurnieden**
Adapted `from/to_bin` to new naming scheme and added `length` as proposed in #348.

### CODE REVIEW

**Comment by czurnieden** [comment]
Left a note in `bn_mp_prime_rand.c` as a reminder that we need to change the type of all length related variables to `size_t` at some point in time. It slowly starts to get a bit messy.

**Comment by minad** [comment]
Doesn't look right. mp_to_xbin should get a maxsize and a written argument.

**Comment by czurnieden** [comment]
> Doesn't look right. mp_to_xbin should get a maxsize and a written argument.

"Boss, we're done! What should we change?" ;-)

I used [@sjaeckel 's  proposition](https://github.com/libtom/libtommath/pull/348#issuecomment-530775486) which was the last state regarding the `mp_x_xbin` functions as far as I am aware, only `mp_to_radix` got an update later.

So you want it differently&mdash;mirroring the behaviour of the `mp_to_radix` if I understood you correctly?
No, problem, the main work (renaming, deprecating the old ones, writing tests, &hellip;) is already done, just do a `s/should/need to/gi` and I'll change it today.

BTW: did you find out what I effed up to cause the `NOP mp_read_write_binaries` message? The tests themselves are good, I tried with `T0(mp_read_write_binaries)`, but it was already late this morning so I had to drop it.

**Comment by czurnieden** [comment]
> did you find out what I effed up to cause the NOP mp_read_write_binaries message?

Well, I did and I was right: I'm an idiot! ;-)

File: bn_mp_to_sbin.c [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
 /* LibTomMath, multiple-precision integer library -- Tom St Denis */
 /* SPDX-License-Identifier: Unlicense */
 
 /* store in signed [big endian] format */
-mp_err mp_to_signed_bin(const mp_int *a, unsigned char *b)
+mp_err mp_to_sbin(const mp_int *a, unsigned char *b, size_t maxlen, size_t *written)
 {
    mp_err err;
-   if ((err = mp_to_unsigned_bin(a, b + 1)) != MP_OKAY) {
+   if ((err = mp_to_ubin(a, b + 1, maxlen, written)) != MP_OKAY) {
```
Comment (minad) [inline bn_mp_to_sbin.c]: maxlen - 1?

File: demo/test.c [hunk 2]
```diff
@@ -2214,6 +2214,70 @@ static int test_s_mp_toom_sqr(void)
    return EXIT_FAILURE;
 }
 
+static int test_mp_read_write_binaries(void)
```
Comment (minad) [inline demo/test.c]: separate tests test_mp_read_write_ubin, test_mp_read_write_sbin

File: bn_mp_to_sbin.c (hunk 1 above)
Comment (czurnieden) [reply bn_mp_to_sbin.c]: > maxlen - 1?

Yeah, added the sign to `written` but forgot to subtract it from the `maxlen`. Will correct.

>  (also check for maxlen==0).

Why? Ah, yes, `maxlen` is `size_t`, right. Will correct.

File: demo/test.c (hunk 2 above)
Comment (czurnieden) [reply demo/test.c]: > minad added this to the v1.2.0 milestone 26 minutes ago

Always that pressure! ;-)

**Comment by minad** [comment]
Status? Could you please squash?

**Comment by czurnieden** [comment]
We are only waiting for your OK, so if you are OK with it, drop me a note and I will squash&finish.

**Comment by minad** [comment]
Looks good

File: bn_mp_to_ubin.c [hunk 3]
```diff
@@ -28,6 +41,9 @@ mp_err mp_to_unsigned_bin(const mp_int *a, unsigned char *b)
    s_mp_reverse(b, x);
```
Comment (sjaeckel) [inline bn_mp_to_ubin.c]: ~and then also change the API of `s_mp_reverse()` to use `size_t`~ that's already done in #348

File: bn_mp_to_ubin.c [hunk 4]
```diff
@@ -1,21 +1,34 @@
 #include "tommath_private.h"
-#ifdef BN_MP_TO_UNSIGNED_BIN_C
+#ifdef BN_MP_TO_UBIN_C
 /* LibTomMath, multiple-precision integer library -- Tom St Denis */
 /* SPDX-License-Identifier: Unlicense */
 
 /* store in unsigned [big endian] format */
-mp_err mp_to_unsigned_bin(const mp_int *a, unsigned char *b)
+mp_err mp_to_ubin(const mp_int *a, unsigned char *b, size_t maxlen, size_t *len)
```
Comment (sjaeckel) [inline bn_mp_to_ubin.c]: `written` instead of `len`

File: bn_mp_to_ubin.c [hunk 5]
```diff
@@ -28,6 +41,9 @@ mp_err mp_to_unsigned_bin(const mp_int *a, unsigned char *b)
    s_mp_reverse(b, x);
    err = MP_OKAY;
 
+   if (len != NULL) {
+      *len = (size_t)x;
```
[the remark this answers is not in the document]
Comment (czurnieden) [reply bn_mp_to_ubin.c]: That's a really good question but I'm pretty sure of the answer: C&P ;-)

**Comment by czurnieden** [comment]
I think I'm doing too many PRs in parallel ;-)

**Comment by minad** [comment]
Squash and rebase?

**Comment by minad** [comment]
@sjaeckel I think this one is ready :+1:

**Pull request merged**